components/brave_vpn/browser/brave_vpn_service.cc

```diff
@@ -774,8 +774,12 @@ void BraveVpnService::SetPurchasedState(
     obs->OnPurchasedStateChanged(state, description);
 
 #if !BUILDFLAG(IS_ANDROID)
-  if (state == PurchasedState::PURCHASED)
+  if (state == PurchasedState::PURCHASED) {
     connection_api_->CheckConnection();
+
+    // Some platform needs to install services to run vpn.
+    connection_api_->MaybeInstallSystemServices();
```

Comment:
`CheckConnection` and `InstallSystemService` are not related operations each other.
`InstallSystemService` could be called before `CheckConnection`. Order doesn't have any meaning.

After startup, we want to know os vpn status if current user is purchased user.
After fetching connection status, we update VPN UI with it.
In `not purchased` state, we don't need to know about os vpn status.
That's why we call `CheckConnection` here.

Reply:
Is that because `MaybeInstallSystemServices` and `InstallBraveWireGuardServiceImpersonated` inside it do not start the service, and `BraveVPNWireguardConnectionAPI::CheckConnection` only checks if the service is running (not whether it's installed)? In any case, I feel a comment in this place of the code would help to clarify why the order does not matter.

Reply:
it's actually the logic above this that seems problematic. We are setting the purchased state before we install the services so there's a race condition here if someone tries to start the vpn before the services are installed. We need a OneShotEvent in `BraveVPNOSConnectionAPI` so any calls to `Connect`/`Disconnect` will be queued until `OnInstallSystemServicesCompleted` is called